From f196a737d830ba7648ff045b8558eb0bcd78e091 Mon Sep 17 00:00:00 2001
From: Kill3rEz <98993017+Kill3rEz@users.noreply.github.com>
Date: Thu, 29 Oct 2026 14:47:29 +0100
Subject: [PATCH] Settings: Add the missing hardware info controllers

hardware_info.xml references controllers for the market name, SoC model and total RAM that do not exist, so the rows stay empty.
---
 res/values/penguin_strings.xml                |  2 +
 .../MarketNamePreferenceController.java       | 47 ++++++++++++++
 .../SoCModelPreferenceController.java         | 56 +++++++++++++++++
 .../TotalRAMPreferenceController.java         | 61 +++++++++++++++++++
 4 files changed, 166 insertions(+)
 create mode 100644 src/com/android/settings/deviceinfo/hardwareinfo/MarketNamePreferenceController.java
 create mode 100644 src/com/android/settings/deviceinfo/hardwareinfo/SoCModelPreferenceController.java
 create mode 100644 src/com/android/settings/deviceinfo/hardwareinfo/TotalRAMPreferenceController.java

diff --git a/res/values/penguin_strings.xml b/res/values/penguin_strings.xml
index 9ffac29ba415..37d40d4b2710 100644
--- a/res/values/penguin_strings.xml
+++ b/res/values/penguin_strings.xml
@@ -7,4 +7,6 @@
     <string name="maintainer_unofficial_build">Unofficial build</string>
     <string name="maintainer_unknown">Unknown</string>
     <string name="penguin_maintainer" translatable="false"></string>
+
+    <string name="penguin_soc_model" translatable="false"></string>
 </resources>
diff --git a/src/com/android/settings/deviceinfo/hardwareinfo/MarketNamePreferenceController.java b/src/com/android/settings/deviceinfo/hardwareinfo/MarketNamePreferenceController.java
new file mode 100644
index 000000000000..b2b338776850
--- /dev/null
+++ b/src/com/android/settings/deviceinfo/hardwareinfo/MarketNamePreferenceController.java
@@ -0,0 +1,47 @@
+/*
+ * Copyright (C) 2026 The PenguinOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.deviceinfo.hardwareinfo;
+
+import android.content.Context;
+import android.os.SystemProperties;
+import android.text.TextUtils;
+
+import com.android.settings.core.BasePreferenceController;
+
+public class MarketNamePreferenceController extends BasePreferenceController {
+
+    private static final String PROP_MARKET_NAME = "ro.product.marketname";
+
+    public MarketNamePreferenceController(Context context, String preferenceKey) {
+        super(context, preferenceKey);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return !TextUtils.isEmpty(getSummary()) ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public boolean useDynamicSliceSummary() {
+        return true;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        return SystemProperties.get(PROP_MARKET_NAME);
+    }
+}
diff --git a/src/com/android/settings/deviceinfo/hardwareinfo/SoCModelPreferenceController.java b/src/com/android/settings/deviceinfo/hardwareinfo/SoCModelPreferenceController.java
new file mode 100644
index 000000000000..75bc0e07e633
--- /dev/null
+++ b/src/com/android/settings/deviceinfo/hardwareinfo/SoCModelPreferenceController.java
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 2026 The PenguinOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.deviceinfo.hardwareinfo;
+
+import android.content.Context;
+import android.os.Build;
+import android.text.TextUtils;
+
+import com.android.settings.R;
+import com.android.settings.core.BasePreferenceController;
+
+public class SoCModelPreferenceController extends BasePreferenceController {
+
+    public SoCModelPreferenceController(Context context, String preferenceKey) {
+        super(context, preferenceKey);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return !TextUtils.isEmpty(getSummary()) ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public boolean useDynamicSliceSummary() {
+        return true;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        final String socModel = mContext.getString(R.string.penguin_soc_model);
+        if (!TextUtils.isEmpty(socModel)) {
+            return socModel;
+        }
+        if (Build.UNKNOWN.equals(Build.SOC_MODEL)) {
+            return null;
+        }
+        if (Build.UNKNOWN.equals(Build.SOC_MANUFACTURER)) {
+            return Build.SOC_MODEL;
+        }
+        return Build.SOC_MANUFACTURER + " " + Build.SOC_MODEL;
+    }
+}
diff --git a/src/com/android/settings/deviceinfo/hardwareinfo/TotalRAMPreferenceController.java b/src/com/android/settings/deviceinfo/hardwareinfo/TotalRAMPreferenceController.java
new file mode 100644
index 000000000000..a37520a4b1e6
--- /dev/null
+++ b/src/com/android/settings/deviceinfo/hardwareinfo/TotalRAMPreferenceController.java
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2026 The PenguinOS Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.settings.deviceinfo.hardwareinfo;
+
+import android.app.ActivityManager;
+import android.content.Context;
+import android.text.format.Formatter;
+
+import com.android.settings.core.BasePreferenceController;
+
+public class TotalRAMPreferenceController extends BasePreferenceController {
+
+    private static final long GIB = 1024L * 1024L * 1024L;
+    private static final long GB = 1000L * 1000L * 1000L;
+    private static final int[] RAM_SIZES_GB = {
+        1, 2, 3, 4, 6, 8, 10, 12, 16, 18, 20, 24, 32, 48, 64
+    };
+
+    public TotalRAMPreferenceController(Context context, String preferenceKey) {
+        super(context, preferenceKey);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return AVAILABLE;
+    }
+
+    @Override
+    public boolean useDynamicSliceSummary() {
+        return true;
+    }
+
+    @Override
+    public CharSequence getSummary() {
+        final ActivityManager.MemoryInfo memoryInfo = new ActivityManager.MemoryInfo();
+        mContext.getSystemService(ActivityManager.class).getMemoryInfo(memoryInfo);
+        final long totalGib = (memoryInfo.totalMem + GIB - 1) / GIB;
+        long sizeGb = totalGib;
+        for (int size : RAM_SIZES_GB) {
+            if (totalGib <= size) {
+                sizeGb = size;
+                break;
+            }
+        }
+        return Formatter.formatShortFileSize(mContext, sizeGb * GB);
+    }
+}
-- 
2.53.0

